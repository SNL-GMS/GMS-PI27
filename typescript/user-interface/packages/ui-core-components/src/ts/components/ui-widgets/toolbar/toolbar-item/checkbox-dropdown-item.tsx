import React from 'react';

import type { CheckboxListProps } from '../../checkbox-list';
import { CheckboxList } from '../../checkbox-list';
import { PopoverButton } from '../../popover-button';
import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isCheckboxDropdownToolbarItem(
  object: unknown
): object is CheckboxDropdownToolbarItemProps {
  return (object as CheckboxDropdownToolbarItemProps).checkboxEnum !== undefined;
}

/**
 * props for {@link CheckboxDropdownToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export type CheckboxDropdownToolbarItemProps = ToolbarItemBase &
  CheckboxListProps & {
    /** callback to onPopUp (list appearing) event */
    onPopUp?(ref?: HTMLDivElement): void;

    /** callback to onPopoverDismissed (list disappears) event */
    onPopoverDismissed?(): void;
  };

/**
 * Represents a dropdown list of checkbox items used within a toolbar
 *
 * @param checkboxItem the checkboxItem to display {@link CheckboxDropdownItem}
 */
export function CheckboxDropdownToolbarItem({
  popoverButtonMap,
  style,
  label,
  tooltip,
  disabled,
  widthPx,
  enumToCheckedMap,
  enumToColorMap,
  checkboxEnum,
  enumKeysToDisplayStrings,
  enumKeysToDividerMap,
  enumKeysToLabelMap,
  onChange,
  onPopUp,
  onPopoverDismissed
}: CheckboxDropdownToolbarItemProps): ToolbarItemElement {
  const handleRef = React.useCallback(
    (ref: PopoverButton): void => {
      if (ref && popoverButtonMap) {
        popoverButtonMap.set(1, ref);
      }
    },
    [popoverButtonMap]
  );

  return (
    <div style={style ?? {}}>
      <PopoverButton
        label={label || ''}
        tooltip={tooltip || ''}
        disabled={disabled}
        popupContent={
          <CheckboxList
            enumToCheckedMap={enumToCheckedMap}
            enumToColorMap={enumToColorMap}
            checkboxEnum={checkboxEnum}
            enumKeysToDisplayStrings={enumKeysToDisplayStrings}
            enumKeysToDividerMap={enumKeysToDividerMap}
            enumKeysToLabelMap={enumKeysToLabelMap}
            onChange={onChange}
          />
        }
        onPopoverDismissed={onPopoverDismissed}
        widthPx={widthPx}
        onClick={onPopUp}
        ref={handleRef}
      />
    </div>
  );
}
