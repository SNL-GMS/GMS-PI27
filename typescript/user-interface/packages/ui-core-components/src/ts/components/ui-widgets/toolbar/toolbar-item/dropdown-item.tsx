import React from 'react';

import type { DropDownProps } from '../../drop-down';
import { DropDown } from '../../drop-down';
import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isDropdownToolbarItem(
  maybeDropdown: unknown
): maybeDropdown is DropdownToolbarItemProps<object> {
  return (maybeDropdown as DropdownToolbarItemProps<object>).dropDownItems !== undefined;
}

/** Internal type to be used as the 2nd type argument for {@link Pick} */
type DropdownPropsForToolbarItem =
  | 'dropDownItems'
  | 'dropDownText'
  | 'disabledDropDownOptions'
  | 'displayLabel'
  | 'value'
  | 'custom'
  | 'onChange';

/**
 * Properties to pass to the {@link DropdownToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export type DropdownToolbarItemProps<T extends object> = ToolbarItemBase &
  Pick<DropDownProps<T>, DropdownPropsForToolbarItem>;

/**
 * Represents a group of static items to display/select within a toolbar
 *
 * @param dropdownItem the dropdownItem to display {@link DropdownItem}
 */
export function DropdownToolbarItem<T extends object>({
  dropDownItems,
  dropDownText,
  disabledDropDownOptions,
  value,
  custom,
  displayLabel,
  onChange,
  style,
  disabled,
  widthPx,
  tooltip,
  label
}: DropdownToolbarItemProps<T>): ToolbarItemElement {
  return (
    <div style={style ?? {}}>
      <DropDown<T>
        onChange={onChange}
        value={value}
        custom={custom}
        dropDownItems={dropDownItems}
        dropDownText={dropDownText}
        disabledDropDownOptions={disabledDropDownOptions}
        disabled={disabled}
        widthPx={widthPx}
        title={tooltip}
        displayLabel={displayLabel}
        label={label}
      />
    </div>
  );
}
