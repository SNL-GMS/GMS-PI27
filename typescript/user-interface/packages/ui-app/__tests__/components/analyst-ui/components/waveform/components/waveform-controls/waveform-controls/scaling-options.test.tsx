import { isCustomToolbarItem } from '@gms/ui-core-components/lib/components/ui-widgets/toolbar/toolbar-item/custom-item';
import { render } from '@testing-library/react';
import * as React from 'react';

import {
  AmplitudeScalingOptions,
  useScalingOptions
} from '../../../../../../../../src/ts/components/analyst-ui/components/waveform/components/waveform-controls/scaling-options';

describe('Scaling Options Toolbar Item', () => {
  it('should exist', () => {
    expect(useScalingOptions).toBeDefined();
  });
  it('should return a ToolbarItem element that matches a snapshot', () => {
    function ScalingOptionWrapper() {
      const newScalingOptionToolbarItem = useScalingOptions(
        AmplitudeScalingOptions.AUTO,
        1,
        () => {
          /* empty */
        },
        () => {
          /* empty */
        },
        'testscaling'
      );
      const { props: itemBase } = newScalingOptionToolbarItem.toolbarItem;
      return isCustomToolbarItem(itemBase) ? itemBase.element : <div />;
    }
    const { container } = render(<ScalingOptionWrapper />);
    expect(container).toMatchSnapshot();
  });
});
