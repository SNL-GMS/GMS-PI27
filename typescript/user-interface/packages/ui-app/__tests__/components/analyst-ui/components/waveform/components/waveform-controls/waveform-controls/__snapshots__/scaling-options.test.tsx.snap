// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scaling Options Toolbar Item should return a ToolbarItem element that matches a snapshot 1`] = `
<div>
  <span
    aria-expanded="false"
    aria-haspopup="true"
    class="bp5-popover-target"
  >
    <button
      class="bp5-button bp5-align-left"
      data-cy="scaling-option-selector"
      title="Waveform amplitude scaling options"
      type="button"
      value="Auto"
    >
      <span
        class="bp5-button-text"
      >
        Scale: 
        Auto
      </span>
      <span
        aria-hidden="true"
        class="bp5-icon bp5-icon-standard bp5-icon-caret-down"
        data-icon="caret-down"
      />
    </button>
  </span>
</div>
`;
